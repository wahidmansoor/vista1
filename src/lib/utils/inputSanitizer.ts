/**
 * üßπ Enhanced Input Sanitization Utilities for AI Agent System
 * 
 * ‚úÖ SECURITY FEATURES:
 * - XSS prevention with comprehensive pattern matching
 * - SQL injection protection
 * - Medical context preservation
 * - Rate limiting validation
 * - Input validation with meaningful error messages
 * 
 * üè• MEDICAL-GRADE SANITIZATION:
 * - Preserves clinical terminology and abbreviations
 * - Handles drug names, measurements, and medical procedures
 * - Maintains formatting for dosage information
 */

export interface SanitizationOptions {
  maxLength?: number;
  allowHtml?: boolean;
  allowSpecialChars?: boolean;
  removeEmojis?: boolean;
  preserveNewlines?: boolean;
  stripUrls?: boolean;
}

export interface ValidationResult {
  isValid: boolean;
  reason?: string;
  sanitizedInput?: string;
}

export interface RateLimitResult {
  allowed: boolean;
  resetTime?: number;
  remainingRequests?: number;
}

export class InputSanitizer {
  private static readonly DEFAULT_MAX_LENGTH = 2000;
  private static readonly MEDICAL_MAX_LENGTH = 5000;
  
  // Enhanced dangerous patterns with medical context awareness
  private static readonly DANGEROUS_PATTERNS = [
    /javascript:/gi,
    /data:/gi,
    /vbscript:/gi,
    /on\w+\s*=/gi, // event handlers like onclick=
    /<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi,
    /<iframe\b[^<]*(?:(?!<\/iframe>)<[^<]*)*<\/iframe>/gi,
    /<object\b[^<]*(?:(?!<\/object>)<[^<]*)*<\/object>/gi,
    /<embed\b[^<]*(?:(?!<\/embed>)<[^<]*)*<\/embed>/gi,
    /<link\b[^<]*>/gi,
    /<meta\b[^<]*>/gi,
    /expression\s*\(/gi, // CSS expression attacks
    /url\s*\(/gi, // CSS url() attacks (but preserve medical URLs if needed)
  ];

  // Medical abbreviations and terms that should be preserved
  private static readonly MEDICAL_TERMS = [
    'mg', 'ml', 'mcg', 'IU', 'bid', 'tid', 'qid', 'prn', 'po', 'iv', 'im', 'sq',
    'CBC', 'CMP', 'LFT', 'BUN', 'Cr', 'eGFR', 'HbA1c', 'PSA', 'CEA', 'CA',
    'ANC', 'WBC', 'RBC', 'Hgb', 'Hct', 'MCV', 'PLT',
    'ECOG', 'Karnofsky', 'AJCC', 'TNM', 'NCCN'
  ];

  /**
   * Enhanced prompt sanitization with medical context preservation
   */
  static sanitizePrompt(input: string, options: SanitizationOptions = {}): string {
    if (!input || typeof input !== 'string') {
      return '';
    }

    const {
      maxLength = this.DEFAULT_MAX_LENGTH,
      allowHtml = false,
      allowSpecialChars = true,
      removeEmojis = false,
      preserveNewlines = true,
      stripUrls = false
    } = options;

    let sanitized = input;

    // Remove null bytes and control characters (preserve medical formatting)
    sanitized = sanitized.replace(/[\x00-\x08\x0B\x0C\x0E-\x1F\x7F]/g, '');

    // Remove dangerous patterns
    this.DANGEROUS_PATTERNS.forEach(pattern => {
      sanitized = sanitized.replace(pattern, '');
    });    // Remove HTML if not allowed
    if (!allowHtml) {
      sanitized = this.stripHtml(sanitized);
    }

    // Handle URLs
    if (stripUrls) {
      // Preserve medical reference URLs but remove others
      sanitized = sanitized.replace(/https?:\/\/(?!.*(?:pubmed|ncbi|nccn|asco|nejm|jco|cancer))[^\s]+/gi, '[URL_REMOVED]');
    }

    // Handle special characters while preserving medical notation
    if (!allowSpecialChars) {
      // Preserve medical notation like Œºg, ¬∞C, %, etc.
      sanitized = sanitized.replace(/[^\w\s\.\,\;\:\!\?\-\+\=\(\)\[\]\{\}\/\%\¬∞\Œº\Œ±\Œ≤\Œ≥\Œ¥]/g, '');
    }

    // Remove emojis if requested
    if (removeEmojis) {
      sanitized = sanitized.replace(/([\u2700-\u27BF]|[\uE000-\uF8FF]|\uD83C[\uDC00-\uDFFF]|\uD83D[\uDC00-\uDFFF]|[\u2011-\u26FF]|\uD83E[\uDD10-\uDDFF])/g, '');
    }

    // Handle newlines
    if (!preserveNewlines) {
      sanitized = sanitized.replace(/\n+/g, ' ');
    } else {
      // Normalize excessive newlines while preserving medical formatting
      sanitized = sanitized.replace(/\n{4,}/g, '\n\n');
    }

    // Remove excessive whitespace but preserve medical spacing
    sanitized = sanitized.replace(/[ \t]+/g, ' ').trim();

    // Truncate to maximum length with intelligent word boundary
    if (sanitized.length > maxLength) {
      sanitized = this.intelligentTruncate(sanitized, maxLength);
    }

    return sanitized;
  }

  /**
   * Sanitizes medical context with higher limits and medical term preservation
   */
  static sanitizeMedicalContext(input: string): string {
    return this.sanitizePrompt(input, {
      maxLength: this.MEDICAL_MAX_LENGTH,
      allowHtml: false,
      stripUrls: false,
      preserveNewlines: true,
      allowSpecialChars: true
    });
  }

  /**
   * Validates prompt content with medical context awareness
   */
  static validatePrompt(input: string): ValidationResult {
    if (!input || typeof input !== 'string') {
      return { isValid: false, reason: 'Input must be a valid string' };
    }

    const sanitized = this.sanitizePrompt(input);
    
    if (sanitized.length === 0) {
      return { isValid: false, reason: 'Prompt cannot be empty after sanitization' };
    }

    if (sanitized.length < 3) {
      return { isValid: false, reason: 'Prompt too short - please provide more detail' };
    }

    // Check for spam-like patterns
    if (/(.)\1{15,}/.test(sanitized)) {
      return { isValid: false, reason: 'Invalid input pattern detected (excessive repetition)' };
    }

    // Check for meaningful content (but allow medical abbreviations)
    const words = sanitized.toLowerCase().split(/\s+/);
    const uniqueWords = new Set(words);
    const medicalTermCount = words.filter(word => 
      this.MEDICAL_TERMS.some(term => word.includes(term.toLowerCase()))
    ).length;

    // More lenient for medical content
    if (words.length > 10 && uniqueWords.size / words.length < 0.2 && medicalTermCount < 2) {
      return { isValid: false, reason: 'Input appears to contain excessive repetition' };
    }

    // Check for potential injection attempts
    const suspiciousPatterns = [
      /union\s+select/gi,
      /drop\s+table/gi,
      /insert\s+into/gi,
      /delete\s+from/gi,
      /update\s+set/gi,
      /exec\s*\(/gi,
      /eval\s*\(/gi
    ];

    for (const pattern of suspiciousPatterns) {
      if (pattern.test(sanitized)) {
        return { isValid: false, reason: 'Potentially malicious content detected' };
      }
    }

    return { isValid: true, sanitizedInput: sanitized };
  }

  /**
   * Enhanced rate limiting with user-specific tracking
   */
  static validateRateLimit(
    userId: string,
    windowMs: number = 60000, // 1 minute
    maxRequests: number = 15 // Higher limit for medical users
  ): RateLimitResult {
    const key = `rate_limit_${userId}`;
    const now = Date.now();
    
    // In production, this should use Redis or a proper database
    const rateLimitStore = (global as any).__rateLimitStore__ || {};
    (global as any).__rateLimitStore__ = rateLimitStore;
    
    const userLimit = rateLimitStore[key] || { count: 0, resetTime: now + windowMs };
    
    if (now > userLimit.resetTime) {
      userLimit.count = 1;
      userLimit.resetTime = now + windowMs;
    } else {
      userLimit.count += 1;
    }
    
    rateLimitStore[key] = userLimit;
    
    return {
      allowed: userLimit.count <= maxRequests,
      resetTime: userLimit.resetTime,
      remainingRequests: Math.max(0, maxRequests - userLimit.count)
    };
  }

  /**
   * Sanitizes module-specific context data
   */
  static sanitizeModuleContext(
    module: string,
    context: Record<string, any>
  ): Record<string, string> {
    const sanitized: Record<string, string> = {};

    for (const [key, value] of Object.entries(context)) {
      if (typeof value === 'string') {
        sanitized[key] = this.sanitizeMedicalContext(value);
      } else if (typeof value === 'number' || typeof value === 'boolean') {
        sanitized[key] = String(value);
      } else if (Array.isArray(value)) {
        sanitized[key] = value
          .filter(item => typeof item === 'string' || typeof item === 'number')
          .map(item => this.sanitizeMedicalContext(String(item)))
          .join(', ');
      } else if (value && typeof value === 'object') {
        // Recursively sanitize nested objects (limit depth to prevent attacks)
        try {
          sanitized[key] = this.sanitizeMedicalContext(JSON.stringify(value));
        } catch {
          sanitized[key] = '[COMPLEX_OBJECT]';
        }
      }
    }

    return sanitized;
  }
  /**
   * Intelligent truncation that preserves medical context
   */
  private static intelligentTruncate(text: string, maxLength: number): string {
    if (text.length <= maxLength) return text;

    let truncated = text.substring(0, maxLength);
    
    // Try to end at a sentence boundary
    const lastPeriod = truncated.lastIndexOf('.');
    const lastQuestion = truncated.lastIndexOf('?');
    const lastExclamation = truncated.lastIndexOf('!');
    const lastSentenceEnd = Math.max(lastPeriod, lastQuestion, lastExclamation);
    
    if (lastSentenceEnd > maxLength * 0.7) {
      truncated = truncated.substring(0, lastSentenceEnd + 1);
    } else {
      // Try to end at a word boundary
      const lastSpace = truncated.lastIndexOf(' ');
      if (lastSpace > maxLength * 0.8) {
        truncated = truncated.substring(0, lastSpace);
      }
    }
    
    return truncated.trim() + '...';
  }

  /**
   * Strips HTML tags from input while preserving medical formatting
   */
  private static stripHtml(input: string): string {
    return input
      .replace(/<[^>]*>/g, '') // Remove HTML tags
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&amp;/g, '&')
      .replace(/&quot;/g, '"')
      .replace(/&#x27;/g, "'")
      .replace(/&#x2F;/g, '/')
      .replace(/&nbsp;/g, ' ');
  }

  /**
   * Sanitizes complete AI request objects
   */
  static sanitizeAIRequest(request: any): any {
    if (!request || typeof request !== 'object') {
      return {};
    }

    const sanitized: any = {};

    // Sanitize common AI request fields
    if (request.prompt) {
      sanitized.prompt = this.sanitizePrompt(request.prompt);
    }

    if (request.context) {
      sanitized.context = this.sanitizeMedicalContext(request.context);
    }

    if (request.history && Array.isArray(request.history)) {
      sanitized.history = request.history
        .filter(item => typeof item === 'string')
        .map(item => this.sanitizeMedicalContext(item));
    }

    if (request.module && typeof request.module === 'string') {
      sanitized.module = request.module.replace(/[^a-zA-Z0-9_-]/g, '');
    }

    if (request.intent && typeof request.intent === 'string') {
      sanitized.intent = request.intent.replace(/[^a-zA-Z0-9_-]/g, '');
    }

    // Copy other safe fields
    ['iterationCount', 'feedbackType', 'mockMode'].forEach(field => {
      if (request[field] !== undefined) {
        sanitized[field] = request[field];
      }
    });

    return sanitized;
  }
}

/**
 * Convenience function for quick prompt sanitization
 */
export function sanitizePrompt(input: string): string {
  return InputSanitizer.sanitizePrompt(input);
}

/**
 * Convenience function for complete request sanitization
 */
export function sanitizeAIRequest(request: any): any {
  return InputSanitizer.sanitizeAIRequest(request);
}
