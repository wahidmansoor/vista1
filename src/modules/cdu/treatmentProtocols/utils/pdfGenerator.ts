import { Protocol, ProtocolDrug } from './types/protocol';

export const generateProtocolPDF = (protocol: Protocol): string => {
  if (protocol.treatment?.drugs?.length) {
    const drugsList = protocol.treatment.drugs.map((drug: ProtocolDrug) => {
      const details = [];
      if (drug.name) details.push(`<strong>${drug.name}</strong>`);
      if (drug.dose) details.push(`Dose: ${drug.dose}`);
      if (drug.administration) details.push(`Route: ${drug.administration}`);
      if (drug.special_notes?.length) {
        details.push(`Notes: ${drug.special_notes.join(', ')}`);
      }
      return details.join(' | ');
    });
    return drugsList.join('\n');
  }
  return '';
};
