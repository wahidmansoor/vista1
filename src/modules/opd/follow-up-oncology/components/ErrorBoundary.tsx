import React, { Component, ErrorInfo, ReactNode } from 'react';

interface Props {
  children: ReactNode;
}

interface State {
  hasError: boolean;
  error?: Error;
}

export class ErrorBoundary extends Component<Props, State> {
  public state: State = {
    hasError: false
  };

  public static getDerivedStateFromError(error: Error): State {
    return { hasError: true, error };
  }

  public componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error('Follow-up module error:', error, errorInfo);
  }

  public render() {
    if (this.state.hasError) {
      return (
        <div className="p-4 bg-red-50 rounded-lg">
          <h4 className="text-red-700 font-medium">Unable to load follow-up guidelines</h4>
          <p className="text-red-600">Please try again or contact support if the problem persists.</p>
        </div>
      );
    }

    return this.props.children;
  }
}
