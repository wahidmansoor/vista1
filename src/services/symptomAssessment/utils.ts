import type { EmergencyCondition } from '../emergency/types';
import type { SymptomSeverity } from './types';
import { mapNumberToSeverity } from '../emergency/config';

export function mapScoreToTriageLevel(score: number): EmergencyCondition['severity'] {
  return mapNumberToSeverity(score);
}

export function calculateSeverityScore(
  baseScore: number,
  factors: {
    duration?: string;
    frequency?: string;
    interference?: number;
  }
): number {
  let score = baseScore;

  // Adjust score based on duration
  if (factors.duration) {
    if (factors.duration.includes('week')) score += 1;
    if (factors.duration.includes('month')) score += 2;
  }

  // Adjust score based on frequency
  if (factors.frequency) {
    if (factors.frequency.includes('daily')) score += 1;
    if (factors.frequency.includes('constant')) score += 2;
  }

  // Adjust score based on interference with daily activities
  if (factors.interference) {
    score += Math.min(factors.interference / 2, 2);
  }

  return Math.min(score, 10); // Cap at 10
}

export function getSeverityDetails(score: number): SymptomSeverity {
  return {
    level: getSymptomSeverityLevel(score),
    score,
    requiresImmediate: score >= 8
  };
}

function getSymptomSeverityLevel(score: number): SymptomSeverity['level'] {
  if (score >= 8) return 'life-threatening';
  if (score >= 6) return 'severe';
  if (score >= 4) return 'moderate';
  return 'mild';
}
