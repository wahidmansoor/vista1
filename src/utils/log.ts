import LogRocket from 'logrocket';
import { getEnvVar } from './environment';

interface ErrorLoggingOptions {
  moduleName?: string;
  retryCount?: number;
  extra?: Record<string, unknown>;
}

/**
 * Centralized error logging function that handles sending errors to multiple destinations
 * Currently supports LogRocket and console.error, can be extended for other services
 */
export function logError(
  error: Error,
  errorInfo?: React.ErrorInfo,
  options: ErrorLoggingOptions = {}
) {
  const { moduleName = 'global', retryCount = 0, extra = {} } = options;

  // Development logging with detailed information
  if (getEnvVar('NODE_ENV') === 'development') {
    console.group('Error logged by logError utility:');
    console.error('Error:', error);
    if (errorInfo) {
      console.error('Component Stack:', errorInfo.componentStack);
    }
    console.error('Module:', moduleName);
    console.error('Retry Count:', retryCount);
    if (Object.keys(extra).length > 0) {
      console.error('Extra Information:', extra);
    }
    console.groupEnd();
  }

  // Send to LogRocket with all available context
  LogRocket.captureException(error, {
    tags: {
      module: moduleName,
      errorName: error.name,
      errorType: error.constructor.name,
      retryCount: retryCount.toString(),
    },
    extra: {
      componentStack: errorInfo?.componentStack || 'No component stack available',
      ...extra
    },
  });

  // Here we can add more logging services as needed:
  // - Sentry.captureException(error, { extra: { moduleName, retryCount, ...extra } });
  // - Send to custom error tracking endpoint
  // - Log to analytics service
}
