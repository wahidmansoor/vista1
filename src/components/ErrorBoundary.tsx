import React, { Component, ErrorInfo, ReactNode } from 'react';
import { AlertOctagon, ArrowLeft, RefreshCcw, ChevronDown, LifeBuoy } from 'lucide-react';
import { logError } from '@/utils/log';

interface Props {
  children?: ReactNode;
  moduleName?: string;
  fallback?: ReactNode;
}

interface State {
  hasError: boolean;
  error: Error | null;
  errorInfo: ErrorInfo | null;
  retryCount: number;
  resetKey: number;
  isDetailsOpen: boolean;
  errorId: string;
}

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      hasError: false,
      error: null,
      errorInfo: null,
      retryCount: 0,
      resetKey: 0,
      isDetailsOpen: false,
      errorId: '',
    };
  }

  static getDerivedStateFromError(error: Error): Partial<State> {
    // Generate a unique error ID for tracking
    const errorId = `error_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    return { hasError: true, error, errorId };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    // Enhanced error information for debugging
    const enhancedErrorInfo = {
      ...errorInfo,
      userAgent: navigator.userAgent,
      url: window.location.href,
      timestamp: new Date().toISOString(),
      stackTrace: error.stack,
      errorId: this.state.errorId,
    };

    this.setState({ error, errorInfo });
    
    // Log error with enhanced context
    logError(error, errorInfo, { 
      retryCount: this.state.retryCount,
      moduleName: this.props.moduleName,
      extra: enhancedErrorInfo
    });

    // Send error to external service if available
    if (window.reportError) {
      window.reportError(this.props.moduleName || 'Unknown', error, enhancedErrorInfo);
    }

    // Console log for immediate debugging
    console.group(`ðŸš¨ Error Boundary Caught Error [${this.state.errorId}]`);
    console.error('Module:', this.props.moduleName);
    console.error('Error:', error);
    console.error('Component Stack:', errorInfo.componentStack);
    console.error('Enhanced Info:', enhancedErrorInfo);
    console.groupEnd();
  }

  handleReset = (): void => {
    this.setState((prevState: State) => ({
      hasError: false,
      error: null,
      errorInfo: null,
      retryCount: prevState.retryCount + 1,
      resetKey: prevState.resetKey + 1,
      isDetailsOpen: false,
      errorId: '',
    }));
  };

  handleNavigateHome = (): void => {
    window.location.href = '/';
  };

  handleToggleDetails = (): void => {
    this.setState((prev: State) => ({
      isDetailsOpen: !prev.isDetailsOpen,
    }));
  };

  renderErrorUI = (): ReactNode => (
    <div style={{
      position: 'fixed',
      top: 0,
      left: 0,
      width: '100vw',
      height: '100vh',
      zIndex: 1300,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      padding: '2rem',
      backgroundColor: 'rgba(255,255,255,0.9)',
      backdropFilter: 'blur(4px)'
    }}>
      <div style={{
        maxWidth: '600px',
        width: '100%',
        padding: '3rem',
        borderRadius: '1rem',
        backgroundColor: 'white',
        boxShadow: '0 25px 50px -12px rgba(0, 0, 0, 0.25)',
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        gap: '1.5rem'
      }}>
        <div style={{
          width: '80px',
          height: '80px',
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          borderRadius: '50%',
          backgroundColor: '#ffebee'
        }}>
          <AlertOctagon size={40} color="#d32f2f" />
        </div>
        
        <h1 style={{
          fontSize: '2rem',
          fontWeight: 700,
          textAlign: 'center',
          margin: 0,
          color: '#1a1a1a'
        }}>
          {this.props.moduleName ? `Error in ${this.props.moduleName}` : 'Something went wrong'}
        </h1>
        
        <p style={{
          color: '#666',
          textAlign: 'center',
          margin: 0,
          fontSize: '1.1rem'
        }}>
          Please try refreshing the page or contact support if the problem persists.
        </p>

        {this.state.error && (
          <div style={{ width: '100%' }}>
            <details>
              <summary 
                style={{
                  display: 'flex',
                  alignItems: 'center',
                  gap: '0.5rem',
                  padding: '0.5rem 1rem',
                  border: '1px solid #ddd',
                  borderRadius: '0.5rem',
                  backgroundColor: 'white',
                  cursor: 'pointer',
                  fontSize: '0.9rem',
                  textTransform: 'none'
                }}
              >
                Technical Details
                <ChevronDown size={16} />
              </summary>
              
              <div 
                data-testid="error-details-box"
                style={{ 
                  marginTop: '1rem',
                  padding: '1rem',
                  backgroundColor: '#f5f5f5',
                  borderRadius: '0.5rem',
                  fontSize: '0.875rem'
                }}
              >
                <div style={{ marginBottom: '1rem' }}>
                  <strong>Error:</strong>
                  <pre style={{ 
                    whiteSpace: 'pre-wrap',
                    margin: '0.5rem 0 0 0',
                    color: '#d32f2f',
                    backgroundColor: 'white',
                    padding: '0.5rem',
                    borderRadius: '0.25rem',
                    border: '1px solid #ffcdd2'
                  }}>
                    {this.state.error.toString()}
                  </pre>
                </div>
                
                {this.state.errorInfo?.componentStack && (
                  <div>
                    <strong>Component Stack:</strong>
                    <pre style={{ 
                      whiteSpace: 'pre-wrap',
                      margin: '0.5rem 0 0 0',
                      fontSize: '0.75rem',
                      backgroundColor: 'white',
                      padding: '0.5rem',
                      borderRadius: '0.25rem',
                      border: '1px solid #e0e0e0',
                      maxHeight: '200px',
                      overflow: 'auto'
                    }}>
                      {this.state.errorInfo.componentStack}
                    </pre>
                  </div>
                )}
              </div>
            </details>
          </div>
        )}

        <div style={{ 
          display: 'flex',
          gap: '1rem',
          marginTop: '1rem',
          flexWrap: 'wrap',
          justifyContent: 'center'
        }}>
          <button 
            onClick={this.handleNavigateHome}
            data-testid="goHomeButton"
            style={{
              display: 'flex',
              alignItems: 'center',
              gap: '0.5rem',
              padding: '0.75rem 1.5rem',
              border: '1px solid #ddd',
              borderRadius: '0.5rem',
              backgroundColor: 'white',
              cursor: 'pointer',
              fontSize: '1rem',
              transition: 'all 0.2s'
            }}
          >
            <ArrowLeft size={18} />
            Go Home
          </button>

          <button 
            onClick={this.handleReset}
            data-testid="tryAgainButton"
            style={{
              display: 'flex',
              alignItems: 'center',
              gap: '0.5rem',
              padding: '0.75rem 1.5rem',
              border: 'none',
              borderRadius: '0.5rem',
              backgroundColor: '#1976d2',
              color: 'white',
              cursor: 'pointer',
              fontSize: '1rem',
              transition: 'all 0.2s'
            }}
          >
            <RefreshCcw size={18} />
            Try Again
          </button>
        </div>

        <a 
          href="mailto:support@example.com"
          style={{
            display: 'flex',
            alignItems: 'center',
            gap: '0.5rem',
            color: '#1976d2',
            textDecoration: 'none',
            fontSize: '0.9rem',
            marginTop: '0.5rem'
          }}
        >
          <LifeBuoy size={18} />
          Contact support
        </a>
      </div>
    </div>
  );

  render(): ReactNode {
    if (this.state.hasError) {
      return this.props.fallback || this.renderErrorUI();
    }

    return (
      <div key={this.state.resetKey} data-testid="recovered">
        {this.props.children}
      </div>
    );
  }
}

export default ErrorBoundary;
