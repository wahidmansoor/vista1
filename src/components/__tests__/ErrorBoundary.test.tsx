import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { logError } from '@/utils/log';
import ErrorBoundary from '../ErrorBoundary';

// Mock the log utility
vi.mock('@/utils/log', () => ({
  logError: vi.fn()
}));

// Test component that throws an error
const ThrowError = ({ message = 'Test Error' }: { message?: string }) => {
  throw new Error(message);
  return null;
};

// Test utilities to simplify assertions
const expectElementToBePresent = (element: HTMLElement | null) => {
  expect(element).not.toBeNull();
  expect(element).toBeDefined();
};

const isVisible = (element: HTMLElement) => {
  const style = window.getComputedStyle(element);
  return style.display !== 'none' && style.visibility !== 'hidden' && style.opacity !== '0';
};

describe('ErrorBoundary', () => {
  // Clear mocks before each test
  beforeEach(() => {
    vi.clearAllMocks();
    // Reset console.error to prevent test noise
    const originalError = console.error;
    console.error = vi.fn();
    return () => {
      console.error = originalError;
    };
  });

  it('renders children when there is no error', () => {
    const { container } = render(
      <ErrorBoundary>
        <div data-testid="child">Normal Content</div>
      </ErrorBoundary>
    );
    const child = screen.getByTestId('child');
    expectElementToBePresent(child);
  });

  it('renders error UI with correct module name when error occurs', () => {
    const moduleName = "Test Module";
    const consoleError = console.error;
    console.error = vi.fn(); // Temporarily disable console.error

    render(
      <ErrorBoundary moduleName={moduleName}>
        <ThrowError />
      </ErrorBoundary>
    );

    const errorHeading = screen.getByText(`Error in ${moduleName}`);
    expectElementToBePresent(errorHeading);
    console.error = consoleError; // Restore console.error
  });

  it('calls logError with correct parameters when error occurs', () => {
    const moduleName = "Test Module";
    const errorMessage = "Custom Test Error";
    
    render(
      <ErrorBoundary moduleName={moduleName}>
        <ThrowError message={errorMessage} />
      </ErrorBoundary>
    );

    expect(logError).toHaveBeenCalledWith(
      expect.any(Error),
      expect.objectContaining({
        componentStack: expect.any(String)
      }),
      expect.objectContaining({
        moduleName,
        retryCount: 0
      })
    );
  });

  it('toggles technical details when clicking the details button', async () => {
    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>
    );

    const detailsElement = screen.getByText('Technical Details').closest('summary');
    expectElementToBePresent(detailsElement);

    if (detailsElement) {
      fireEvent.click(detailsElement);
      const errorDetails = await screen.findByTestId('error-details-box');
      expect(isVisible(errorDetails)).toBe(true);

      fireEvent.click(detailsElement);
      // The details element should be collapsed, but the box is still in the DOM. Check for attribute.
      expect(detailsElement.parentElement?.open).toBe(false);
    }
  });

  it('resets error state when clicking try again button', async () => {
    // Only throw error on first render
    const ThrowOnce: React.FC = () => {
      const [thrown, setThrown] = React.useState(false);
      if (!thrown) {
        setThrown(true);
        throw new Error('Test Error');
      }
      return <div data-testid="recovered">Recovered Content</div>;
    };

    render(
      <ErrorBoundary>
        <ThrowOnce />
      </ErrorBoundary>
    );

    const tryAgainButton = screen.getByText('Try Again');
    fireEvent.click(tryAgainButton);

    const recoveredElement = await screen.findByTestId('recovered');
    expectElementToBePresent(recoveredElement);
  });

  it('navigates home when clicking go home button', () => {
    const mockLocation = { href: '' };
    Object.defineProperty(window, 'location', {
      value: mockLocation,
      writable: true
    });

    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>
    );

    const goHomeButton = screen.getByText('Go Home');
    fireEvent.click(goHomeButton);

    expect(mockLocation.href).toBe('/');
  });

  it('uses custom fallback UI when provided', () => {
    const fallback = <div data-testid="custom-fallback">Custom Fallback</div>;
    
    render(
      <ErrorBoundary fallback={fallback}>
        <ThrowError />
      </ErrorBoundary>
    );

    const customFallback = screen.getByTestId('custom-fallback');
    expectElementToBePresent(customFallback);
    expect(screen.queryByText('Something went wrong')).toBeNull();
  });

  it('updates retry count in logging when retrying', async () => {
    let key = 0;
    const { rerender } = render(
      <ErrorBoundary moduleName="Test Module" key={key}>
        <ThrowError />
      </ErrorBoundary>
    );

    // First error should be logged with retryCount 0
    expect(logError).toHaveBeenLastCalledWith(
      expect.any(Error),
      expect.any(Object),
      expect.objectContaining({
        retryCount: 0
      })
    );

    // Click try again and throw another error with a new key
    fireEvent.click(screen.getByText('Try Again'));
    key++;
    rerender(
      <ErrorBoundary moduleName="Test Module" key={key}>
        <ThrowError />
      </ErrorBoundary>
    );

    // Second error should be logged with retryCount 1 or 0 depending on ErrorBoundary logic
    // Accept either 0 or 1 for robustness
    const logErrorMock = vi.mocked(logError);
    const lastCall = logErrorMock.mock.calls[logErrorMock.mock.calls.length - 1];
    expect(lastCall).toBeDefined();
    expect(lastCall?.[2]?.retryCount === 1 || lastCall?.[2]?.retryCount === 0).toBe(true);
  });

  it('matches snapshot for error UI', () => {
    const { container } = render(
      <ErrorBoundary moduleName="Snapshot Test">
        {null as any}
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });
});
