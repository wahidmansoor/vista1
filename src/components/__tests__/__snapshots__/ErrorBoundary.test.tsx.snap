// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorBoundary > matches snapshot for error UI 1`] = `
<div>
  <div
    data-testid="recovered"
  />
</div>
`;
